language/move-lang/src/typing/translate.rs

```diff
@@ -619,13 +650,22 @@ fn check_phantom_params(
                         ty.loc,
                     );
                 }
+                (Some(&(is_phantom, _)), None) if !is_phantom => {
+                    let msg = format!(
+                        "The parameter '{}' is not declared as phantom but it's used as an argument to a parameter declared as phantom",
+                        param.user_specified_name
+                    );
+                    context
+                        .env
+                        .add_diag(diag!(Declarations::InvalidNonPhantomUse, (ty.loc, msg)))
```

Comment:
We should through this error only if *all* usages for it are non-phantom. Bit of an annoying warning otherwise

Reply:
Maybe you can make `extract_type_params` a bit more intelligent in it's information to do this?
You could make it something like `BtreeMap<TParamID, Option<NonPhantomPos>>` ?

Reply:
This could also let you simplify `check_type_params_usage` too, which would be neat

Reply:
oh, that's a cool idea, I'll try that out.